tests/pipeline_test.py

```diff
@@ -388,3 +388,98 @@ def provide_C(x: A[T1], y: B[T2]) -> C[T1, T2]:
         [provide_int, provide_float, provide_A, provide_B, provide_C]
     )
     assert pipeline.compute(C[int, float]) == C[int, float](1, 2.0)
+
+
+def test_instance_provider() -> None:
+    Result = NewType('Result', float)
+
+    def f(x: int, y: float) -> Result:
+        return Result(x / y)
+
+    pl = sl.Pipeline([f])
+    pl[int] = 3
+    pl[float] = 2.0
+    assert pl.compute(int) == 3
+    assert pl.compute(float) == 2.0
+    assert pl.compute(Result) == 1.5
+
+
+def test_provider_NewType_instance() -> None:
+    A = NewType('A', int)
+    pl = sl.Pipeline([])
+    pl[A] = A(3)
+    assert pl.compute(A) == 3
+
```

Comment:
Maybe this kind of test case can be useful...?

```suggestion

def test_provider_Newtype_instance_same_supertype() -> None:
    pl = sl.Pipeline()
    Str = NewType('Str', str)
    IntStr = NewType('IntStr', str)
    pl[IntStr] = Str('0')
    assert pl.compute(IntStr) == '0'

```

Reply:
I suppose this is just a "limitation" of how `NewType` works. I think the type checker should be able to catch this?

Reply:
`mypy` does not, apparently. :(

Reply:
This one is not caught by `mypy` or `pyre` either.
`TypeVar` seems quite loose if it's not used in `Generic` ...
If your intention is not allowing this at all, maybe adding it to the document should be sufficient...?

```
from typing import NewType, Dict, Type, TypeVar, Any
A = NewType('A', str)
B = NewType('B', str)

T = TypeVar("T")

class C:
    def __init__(self) -> None:
        self.type_dict: Dict[Any, Any] = dict()
    
    def __setitem__(self, tp: Type[T], _object: T) -> None:
        self.type_dict[tp] = _object

c = C()
c[A] = B('str')
c[int] = '0'
```

Reply:
> If your intention is not allowing this at all, maybe adding it to the document should be sufficient...?

Not allowing what?

Reply:
To do this, setting an item of type `Str` as a provider of `IntStr` in this example.
```
 pl = sl.Pipeline()
Str = NewType('Str', str)
IntStr = NewType('IntStr', str)
pl[IntStr] = Str('0')
assert pl.compute(IntStr) == '0'
```

Reply:
Well, I don't see how I could prevent that, except by disallowing `NewType` as key?